### PR TITLE
Add UpgradeBehavior deny

### PR DESCRIPTION
**Pull request by Trenly**
- [x] I have signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs).
- [x] This pull request is related to an issue.
  - #3510
  - #3482

-----

 ###### Microsoft Reviewers: codeflow:open?pullrequest=https://github.com/microsoft/winget-cli/pull/3512&drop=dogfoodAlpha

### CODE REVIEW

**Comment by denelon** [comment]
We would have the work in the client and the schema here, as well as the REST schema and the docs at winget-pkgs.

**Comment by Trenly** [comment]
cc @ryfu-msft

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 1]
```diff
@@ -1694,6 +1694,9 @@ Please specify one of them using the --source option to proceed.</value>
     <value>{0} package(s) have a pin that needs to be removed before upgrade</value>
     <comment>{Locked="{0}"} {0} is a placeholder that is replaced by an integer number of packages with pins that prevent upgrade</comment>
   </data>
+  <data name="UpgradeBlockedByManifest" xml:space="preserve">
+    <value>The package cannot be upgraded using winget. Please use the method provided by the publisher for updating this package.</value>
```
Comment (yao-msft) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1698]:
```suggestion
    <value>The package cannot be upgraded using winget. Please use the method provided by the publisher for upgrading this package.</value>

```

nit: use consistent terms

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp [hunk 2]
```diff
@@ -594,6 +594,16 @@ namespace AppInstaller::CLI::Workflow
             AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_INSTALLER_PROHIBITS_ELEVATION);
         }
 
+        // This installer cannot be used to upgrade the currently installed application
+        // Because the upgrade mechanism may be package-specific, simply block.
+        bool isUpdate = WI_IsFlagSet(context.GetFlags(), Execution::ContextFlag::InstallerExecutionUseUpdate);
+        UpdateBehaviorEnum updateBehavior = context.Get<Execution::Data::Installer>().value().UpdateBehavior;
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/InstallFlow.cpp:600]:
```suggestion
        UpdateBehaviorEnum updateBehavior = installer->UpdateBehavior;

```

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp [hunk 3]
```diff
@@ -594,6 +594,16 @@ namespace AppInstaller::CLI::Workflow
             AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_INSTALLER_PROHIBITS_ELEVATION);
         }
 
+        // This installer cannot be used to upgrade the currently installed application
+        // Because the upgrade mechanism may be package-specific, simply block.
+        bool isUpdate = WI_IsFlagSet(context.GetFlags(), Execution::ContextFlag::InstallerExecutionUseUpdate);
+        UpdateBehaviorEnum updateBehavior = context.Get<Execution::Data::Installer>().value().UpdateBehavior;
+        if (isUpdate && (updateBehavior == UpdateBehaviorEnum::Deny)) {
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/InstallFlow.cpp:601]:
```suggestion
        if (isUpdate && (updateBehavior == UpdateBehaviorEnum::Deny))
        {
```

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp [hunk 4]
```diff
@@ -594,6 +594,16 @@ namespace AppInstaller::CLI::Workflow
             AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_INSTALLER_PROHIBITS_ELEVATION);
         }
 
+        // This installer cannot be used to upgrade the currently installed application
+        // Because the upgrade mechanism may be package-specific, simply block.
+        bool isUpdate = WI_IsFlagSet(context.GetFlags(), Execution::ContextFlag::InstallerExecutionUseUpdate);
+        UpdateBehaviorEnum updateBehavior = context.Get<Execution::Data::Installer>().value().UpdateBehavior;
+        if (isUpdate && (updateBehavior == UpdateBehaviorEnum::Deny)) {
+            AICLI_LOG(CLI, Info, << "Manifest specifies update behavior is denied. The attempt will be cancelled.");
+            context.Reporter.Error() << Resource::String::UpgradeBlockedByManifest() << std::endl;
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/InstallFlow.cpp:603]:
```suggestion
            context.Reporter.Error() << Resource::String::UpgradeBlockedByManifest << std::endl;

```

**Review by yao-msft — DISMISSED** [review]

**Comment by yao-msft** [comment]
There are a couple e2e tests related to portable failed because of this new addition of portable 4.0.0.0 manifest with upgrade deny.

We should probably create a new package identifier (2 manifests) like AppInstallerTest.TestUpgradeDeny so it's isolated and not affecting other tests.

**Comment by Trenly** [comment]
Makes sense; I'll take a look

**Comment by Trenly** [comment]
@ryfu-msft - I took a look at the pester tests and I don't see any reason why the BuildPowerShellModule tests would be failing, especially when I didn't touch anything about the PowerShell modules or COM, and it seems like the pipelines are showing a COM error. Unless the changes to the SharedLib files are interfering somehow?

>  COMException: Failed to create instance: -2147023174

**Comment by ryfu-msft** [comment]
> @ryfu-msft - I took a look at the pester tests and I don't see any reason why the BuildPowerShellModule tests would be failing, especially when I didn't touch anything about the PowerShell modules or COM, and it seems like the pipelines are showing a COM error. Unless the changes to the SharedLib files are interfering somehow?
> 
> > COMException: Failed to create instance: -2147023174

This seems like a transient error, lets see if the rerun fixes it. You must be lucky today 😄

**Pull request merged**